app/engine/Flywheel.js

```diff
@@ -0,0 +1,337 @@
+
+  The calculation of angular velocity and acceleration is based on Quadratic Regression, as the second derivative tends to be
+  quite fragile when small errors are thrown in the mix. The math behind this approach can be found in https://physics.info/motion-equations/
+  which is intended for simple linear motion, but the formula are identical when applied to angular distances, velocities and
+  accelerations.
+*/
+
+import loglevel from 'loglevel'
+import { createStreamFilter } from './utils/StreamFilter.js'
+import { createSeries } from './utils/Series.js'
+import { createOLSLinearSeries } from './utils/OLSLinearSeries.js'
+import { createTSQuadraticSeries } from './utils/FullTSQuadraticSeries.js'
+const log = loglevel.getLogger('RowingEngine')
+
+function createFlywheel (rowerSettings) {
+  const angularDisplacementPerImpulse = (2.0 * Math.PI) / rowerSettings.numOfImpulsesPerRevolution
+  const flankLength = Math.max(3, rowerSettings.flankLength)
+  const minimumDragFactorSamples = Math.floor(rowerSettings.minimumRecoveryTime / rowerSettings.maximumTimeBetweenImpulses)
+  const minumumTorqueBeforeStroke = rowerSettings.minumumForceBeforeStroke * (rowerSettings.sprocketRadius / 100)
+  const currentDt = createStreamFilter(rowerSettings.smoothing, rowerSettings.maximumTimeBetweenImpulses)
+  const _deltaTime = createOLSLinearSeries(flankLength)
+  const _angularDistance = createTSQuadraticSeries(flankLength)
+  const _angularVelocityMatrix = []
+  const _angularAccelerationMatrix = []
+  const drag = createStreamFilter(rowerSettings.dragFactorSmoothing, (rowerSettings.dragFactor / 1000000))
+  const recoveryDeltaTime = createOLSLinearSeries()
+  const strokedetectionMinimalGoodnessOfFit = rowerSettings.minimumStrokeQuality
+  const minumumRecoverySlope = createStreamFilter(rowerSettings.dragFactorSmoothing, rowerSettings.minumumRecoverySlope)
+  let _deltaTimeBeforeFlank
+  let _angularVelocityAtBeginFlank
+  let _angularVelocityBeforeFlank
+  let _angularAccelerationAtBeginFlank
+  let _angularAccelerationBeforeFlank
+  let _torqueAtBeginFlank
+  let _torqueBeforeFlank
+  let lastKnownGoodDatapoint
+  let inRecoveryPhase
+  let maintainMetrics
+  let totalNumberOfImpulses
+  let totalTimeSpinning
+  let currentCleanTime
+  let currentRawTime
+  let currentAngularDistance
+  reset()
+  maintainMetrics = true
+
+  function pushValue (dataPoint) {
+    if (dataPoint <= rowerSettings.minimumTimeBetweenImpulses || dataPoint > (1.7 * lastKnownGoodDatapoint)) {
+      // Remove the completely deviating currentDt's, which typically are found after a long pause
+      // TODO: Omgaan met waarden boven 2 * rowerSettings.maximumTimeBetweenImpulses
+      // TODO: Omgaan met waarden boven 2 * vorige waarde --> extra impuls er tussen!
+      dataPoint = 1.7 * lastKnownGoodDatapoint
+    }
+
+    currentDt.push(dataPoint)
+    lastKnownGoodDatapoint = currentDt.clean()
```

Comment:
> Here you are using the median of the raw delta times to update the last known good data point.

This is the running median filter, which by default is set to 1 (i.e. no filtering at all).

> I think this is not necessarily the best approach: What we know is that delta times are either increasing or decreasing (i.e. the current delta time is either higher or lower than the last one). This should mean that the median would never fit the curve since that is a value from a few rotations ago.

Could you clarify "a few rotations ago"? This is a standard approach for signal processing. As it is a **running** median filter, with maximum size after it will clear out data, there is absolutely no reason for data to become stale. By default is set to 1 (effectively turning it off as the median of one value is the same value) and where in extreme cases with lots of noise (see the setup procedure) it can be set to 3 or 5.

I've tested this filter extremely intensively when the TS-regression wasn't optimal yet, and the practical effect is that a **running** median filter of 3 kills all extreme outliers, but typically in the datastream the clean CurrentDt values just lag one impulse behind the raw CurrentDt values. So in practice the datastream reamins intact, it just gets systematically delayed a bit (and with extremes removed).

When set to 5, the **running** median filter will lag two impulses but will still move in the same direction as the data (thus an upward CurrentDt's make it go upward, a downward CurrentDt will make it go downward). In essence, it will delay the data stream, but it will not hurt it. A side effect of this filter is that it will make sharp changes in currentDt more blunt (as there is no way to ditinguish between a real peak or a noise spike).

The last raw datapoint has the huge disadvantage that extremely likely will be noise. The typical issue is that a datapoint gets a delayed read (making that currentDt too long), followed by a normal read (but as the previous one was too long, its currentDt will be shortened by this). So you get one too long reading followed by a too extreme short one. Looking at all the data from a NordicTrack and a Concept2, 90%-95% of the noise is of that type. This is an important realisation, as any practical type of polynomial regression requires the noise to be random and symmetric, allowing us to use regression in the first place.

So copying the raw currentDt directly from the previous value would introduce more noise. Like I said, the argument can be made that it is dead code for most machines, but I'm open for better solutions here.

Reply:
> > Here you are using the median of the raw delta times to update the last known good data point.
> 
> This is the running median filter, which by default is set to 1 (i.e. no filtering at all).

Than that is irrelevant that is true (although I thought it is maxed at 3 -> Math.max(3, rowerSettings.flankLength))

> 
> > I think this is not necessarily the best approach: What we know is that delta times are either increasing or decreasing (i.e. the current delta time is either higher or lower than the last one). This should mean that the median would never fit the curve since that is a value from a few rotations ago.
> 
> Could you clarify "a few rotations ago"? This is a standard approach for signal processing. As it is a **running** median filter, with maximum size after it will clear out data, there is absolutely no reason for data to become stale. By default is set to 1 (effectively turning it off as the median of one value is the same value) and where in extreme cases with lots of noise (see the setup procedure) it can be set to 3 or 5.

I might be missing something there but this is what I meant by few rotations ago:
Lets say these are the raw impulse delta times in microsecond 
40518
38724
37514
36138
34905
34119
11020
32079
31491
8201

If the max size is 3, than of course the median will be always the one before the last element (i.e. in case of the follwing: 32079, 31491, 8201-> 31491; or in case 36138, 34905, 34119). That is fine it fits the curve. Actually it means the same what I meant by using the "last valid delta time".

But if you have 5 for instance then you would end up with the following: 37514, 36138, 34905, 34119, 11020 -> 34905. Where the 34905 is bigger than the 34119 i.e. it does not fit the curve.

> 
> I've tested this filter extremely intensively when the TS-regression wasn't optimal yet, and the practical effect is that a **running** median filter of 3 kills all extreme outliers, but typically in the datastream the clean CurrentDt values just lag one impulse behind the raw CurrentDt values. So in practice the datastream reamins intact, it just gets systematically delayed a bit (and with extremes removed).
> 
> When set to 5, the **running** median filter will lag two impulses but will still move in the same direction as the data (thus an upward CurrentDt's make it go upward, a downward CurrentDt will make it go downward). In essence, it will delay the data stream, but it will not hurt it. A side effect of this filter is that it will make sharp changes in currentDt more blunt (as there is no way to ditinguish between a real peak or a noise spike).
> 
> The last raw datapoint has the huge disadvantage that extremely likely will be noise. The typical issue is that a datapoint gets a delayed read (making that currentDt too long), followed by a normal read (but as the previous one was too long, its currentDt will be shortened by this). So you get one too long reading followed by a too extreme short one. Looking at all the data from a NordicTrack and a Concept2, 90%-95% of the noise is of that type. This is an important realisation, as any practical type of polynomial regression requires the noise to be random and symmetric, allowing us to use regression in the first place.

I meant to use the last clean data point not the raw one (by accepted I meant its a clean one that fits the curve).

> 
> So copying the raw currentDt directly from the previous value would introduce more noise. Like I said, the argument can be made that it is dead code for most machines, but I'm open for better solutions here.

So considering my above explanation using the last correct value would be clearer intent as effectively with a max of 3 we derive to the same thing.
Lets say we have the following DTs:

40518
38724
37514
36138
34905
34119
8201

Instead of the 8201, we would use the 34119 that we know is clean and does not interfere with the curve (I think:))

Reply:
> > This is the running median filter, which by default is set to 1 (i.e. no filtering at all).
> Than that is irrelevant that is true (although I thought it is maxed at 3 -> Math.max(3, rowerSettings.flankLength))

I can imagine you thinking that, as that part of the code is filled with settings. I also had to double check as a mistake is easily made there. The initialisation is:

```js
const currentDt = createStreamFilter(rowerSettings.smoothing, rowerSettings.maximumTimeBetweenImpulses)
```

where the first value determines the length of the running median filter, and the second (optional) value the default value when the value is called when empty. The default config is:

```js
// Smoothing determines the length of the running average for filtering the currentDt, 1 effectively turns it off
smoothing: 1,
```

None of the current known configurations need it, based on their test data. However, there are quite a lot of undocumented machines out there, so removing it was a bit too dangerous for my taste as I don't know what issues people are fighting.

So the manual reads:

> smoothing is the ultimate fallback mechanism for rowers with very noisy data. For all known rowers currently maintained by Open Rowing Monitor, **NONE** needed this, so only start working with this when the raw files show you have a very noisy signal, physical measures don't work and you can't get your stroke detection to work with other means (please note that we design the mechanisms here to be robust, so they can take a hit).
> This is a running median filter, effectively killing any extreme values. By default, it is set to 1 (off). A value of 3 will allow it to completely ignore any single extreme values, which should do the trick for most rowers.